File: common/src/main/java/org/apache/celeborn/common/client/MasterEndpointResolver.java
package org.apache.celeborn.common.client;

import org.apache.celeborn.common.CelebornConf;
import org.apache.celeborn.common.protocol.RpcNameConstants;

import java.util.Iterator;
import java.util.List;

public abstract class MasterEndpointResolver {
    private final CelebornConf conf;
    private final boolean isWorker;
    private String masterEndpointName;
    protected List<String> activeMasterEndpoints;
    protected int currentIndex;

    public MasterEndpointResolver(CelebornConf conf, boolean isWorker) {
        this.conf = conf;
        this.isWorker = isWorker;

        init();
    }

    private void init() {
        if (isWorker && conf.internalPortEnabled()) {
            // For worker, we should use the internal endpoints if internal port is enabled.
            this.masterEndpointName = RpcNameConstants.MASTER_INTERNAL_EP;
            resolve(conf.masterInternalEndpoints());
        } else {
            this.masterEndpointName = RpcNameConstants.MASTER_EP;
            resolve(conf.masterEndpoints());
        }
    }

    public String getMasterEndpointName() {
        return masterEndpointName;
    }

    public List<String> getActiveMasterEndpoints() {
        return activeMasterEndpoints;
    }

    abstract protected void resolve(String[] endpoints);

    abstract protected void update(String[] endpoints);
}

File: common/src/main/scala/org/apache/celeborn/common/client/MasterEndpointResolver.scala
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.celeborn.common.client

import java.util.concurrent.atomic.AtomicBoolean

import scala.collection.JavaConverters._

import org.apache.celeborn.common.CelebornConf
import org.apache.celeborn.common.internal.Logging
import org.apache.celeborn.common.protocol.RpcNameConstants

abstract class MasterEndpointResolver(
    private val conf: CelebornConf,
    private val isWorker: Boolean) extends Logging {

  private val masterEndpointName: String = {
    if (isWorker && conf.internalPortEnabled) {
      // For worker, we should use the internal endpoints if internal port is enabled.
      resolve(conf.masterInternalEndpoints)
      RpcNameConstants.MASTER_INTERNAL_EP
    } else {
      resolve(conf.masterEndpoints)
      RpcNameConstants.MASTER_EP
    }
  }

  protected val updated = new AtomicBoolean(false)

  def getMasterEndpointName: String = masterEndpointName.get

  def getActiveMasterEndpoints: java.util.List[String] = activeMasterEndpoints.get.asJava

  def getUpdatedAndReset(): Boolean = updated.compareAndSet(true, false)

  protected def resolve(endpoints: Array[String]): Unit

  protected def update(endpoints: Array[String]): Unit
}
